### PR TITLE
Add ipynb conversion

### PR DESCRIPTION
**Pull request by FabienRoger**
This is a draft, many required features are missing.
What is there
1. A basic converter which uses the default code block and fills it with the pieces of codes, and align them left to right
2. The load button now supports loading a ipynb file
What is missing
1. Many crucial things (such as markdown and outputs) are ignored by the converter (see the TODOs in the python file)
2. There is not dedicated "load ipynb button" for now. It works with the load button when a .ipynb file is selected.
3. Unit tests might be required for this one, but are not yet present

### CODE REVIEW

File: opencodeblocks/scene/ipynb_conversion.py
Comment (vanyle) [inline opencodeblocks/scene/ipynb_conversion.py:49]: Less TODO comments. That's why there are github comments and issues. One is fine but there are too many.

File: opencodeblocks/scene/ipynb_conversion.py
Comment (vanyle) [inline opencodeblocks/scene/ipynb_conversion.py:41]: Set the height / width of data to match the number of line of the source to avoid scrolling when reading the imported file.

File: opencodeblocks/scene/ipynb_conversion.py
Comment (vanyle) [inline opencodeblocks/scene/ipynb_conversion.py:37]: The empty block might change places in the future. The data fields should be set from scratch.
```python
data = OrderedDict([
  ("id",0),
  # ...
])
```
You can checkout the `serialize` in `block.py` as an example

File: opencodeblocks/scene/scene.py
Comment (vanyle) [inline opencodeblocks/scene/scene.py:165]: This function is the same as `load_from_ipyg`. Consider renaming `load_from_ipyg` to something like `load_from_json`.

**Review by vanyle — CHANGES_REQUESTED** [review]

File: opencodeblocks/scene/ipynb_conversion.py
Comment (vanyle) [reply opencodeblocks/scene/ipynb_conversion.py:37]: 27022fbe2f4960fcc0c23b90b8ec147a8c2c060d
You can do something similar to what you did in this commit.

Comment (vanyle) [reply opencodeblocks/scene/ipynb_conversion.py:37]: In the future, the serialization process will be more robust and we'll be able to not set some fields like splitter and enjoy sane defaults.

**Comment by FabienRoger** [comment]
Update on the state of the pull request

What is there
1. A basic converter which add pieces of codes, and align them left to right, and align markdown top to bottom
2. The load button now supports loading a ipynb file
3. Automatic resizing to see the hole code

Here is what it looks like 
image

What is missing
1. Many crucial things (such outputs) are ignored by the converter
2. There is not dedicated "load ipynb button" for now. It works with the load button when a .ipynb file is selected.
3. Unit tests might be required for this one, but are not yet present
4. Titles ?
5. Automatic text size detection (here the expected text size is fixed)
6. Make it look better ?

**Comment by vanyle** [comment]
You're right about titles. I guess it depends on your use of markdown inside jupyter notebooks.
One solution would be to merge 2 jupyter notebook cells into one when the markdown cell is short and only contains one line,
but to keep the 2 cell separate if there are multiple lines or if the cell uses markdown specific syntax like `#`, `*italics*` or `\`quotes\``

**Comment by FabienRoger** [comment]
Sockets are added and I have implemented your suggestion ("One solution would be to merge 2 jupyter notebook cells into one when the markdown cell is short")

**Comment by FabienRoger** [comment]
What is missing and will be added in future branches :
1. Conversion from ipyg to ipynb
2. Display outputs saved in the ipynb file
3. Unit tests
4. Automatic text size detection (here the expected text size is fixed)
5. A dedicated "load ipynb button" for now. It works with the load button when a .ipynb file is selected. Should it really be added ?

**Comment by FabienRoger** [comment]
I need help with the following:
1. How to generate proper ids ? I don't find the function that does it.
2. How to get the font size (for the moment, it just uses "12" when converting from ipynb to ipyg)

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:29]:
```suggestion
```
id should not be present to be generated

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:26]:
```suggestion
    edges_data: List[OrderedDict] = get_edges_data(data, blocks)
```

`add_sockets` should be named `get_edges_data` to match `get_blocks_data`
Also original data should be passed to `get_edges_data` in case edges data is present in .ipynb metadata

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:25]:
```suggestion
    blocks_data: List[OrderedDict] = get_blocks_data(data)
```

blocks_data should be used instead of blocks. blocks should referer to actual OCBBlock instances

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:241]: Unused function
```suggestion
```

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:201]: Should not add random metadata, deserialization should handle missing arguments with default ones. If that's not the case, raise an issue and start a bugfix branch from master.
```suggestion
def get_default_input_socket(socket_id: int) -> OrderedDict:
    """Returns the default input socket with the corresponding id"""
    return {"id": socket_id, "type": "input"}
```

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:219]: Should not add random metadata, deserialization should handle missing arguments with default ones. If that's not the case, raise an issue and start a bugfix branch from master.
```suggestion
def get_default_output_socket(socket_id: int, block_width: int) -> OrderedDict:
    """Returns the default output socket with the corresponding id"""
    return {"id": socket_id, "type": "output"}
```

File: opencodeblocks/scene/ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/ipynb_conversion.py:234]: `path_type` should be handled by deserialization

```suggestion
def get_default_edge(
    edge_id: int,
    edge_start_block_id: int,
    edge_start_socket_id: int,
    edge_end_block_id: int,
    edge_end_socket_id: int,
) -> OrderedDict:
    return {
        "id": edge_id,
        "source": {"block": edge_start_block_id, "socket": {"id": socket_id, "type": "output"},
        "destination": {"block": edge_end_block_id, "socket": {"id": socket_id, "type": "output"}},
    }
```

**Review by MathisFederico — CHANGES_REQUESTED** [review]
A first review !
Thanks for the contribution !

**Comment by MathisFederico** [comment]
>     1. How to generate proper ids ? I don't find the function that does it.

You can use a generator or just select a number of unique random high enough integers.
For example, for edge, blocks and sockets id, you just need to gather 5*len(blocks) unique integers and you will be good to go !
We will probably refactor id-managment later on anyway.

>     2. How to get the font size (for the moment, it just uses "12" when converting from ipynb to ipyg)

Do you mean how to set the font size ? For source text or/and titles ? If yes you can add it in metadata.
Not so sure why you care about font size for this .ipynb conversion.

**Comment by FabienRoger** [comment]
Nice suggestions ! I added them.

> Do you mean how to set the font size ? For source text or/and titles ? If yes you can add it in metadata.
> Not so sure why you care about font size for this .ipynb conversion.
Blocks width and heights are chosen such that the code fits. I need the text size (or a better solution) to find how much space the code takes.

**Comment by MathisFederico** [comment]
> Blocks width and heights are chosen such that the code fits. I need the text size (or a better solution) to find how much space the code takes.

I see ! It's not that important for now, we will improve this in a future branch, just choose a 'good enough value' for now.

**Review by MathisFederico — APPROVED** [review]
I would not have thought of it like this, but I like this Serializable refactor !
Let's wait for @vanyle approval before merging anything, but it looks good to me !

File: opencodeblocks/scene/from_ipynb_conversion.py
Comment (MathisFederico) [inline opencodeblocks/scene/from_ipynb_conversion.py:12]: I still think this should be edges_data though
```suggestion
    edges_data: List[OrderedDict] = get_edges_data(blocks_data)
```

**Review by MathisFederico — CHANGES_REQUESTED** [review]

**Comment by vanyle** [comment]
image
The font size seems to vary from computer to computer. For example, on my machine, the height of the block needs to be about doubled.

We cannot hardcode `TEXT_SIZE_TO_HEIGHT_RATIO` into a constant, it needs to be computed.

Look into the `QFontMetrics(font)` object. Given a string, this object can return the bounding box of said string.
The font can be retreived from the `PythonEditor` class (inside `pyeditor.py`).

**Comment by vanyle** [comment]
Apart from that, I would also like to say that putting all the default values for cells serialization in their own file was a great idea !

**Review by vanyle — DISMISSED** [review]
Code is very clean !
Just look at the comment for the remaining issue. You can merge it anyway if you want. Because the issue is specific to my machine, it might be better if I focus on fixing it as ratios between real height and height inside a QGraphicsView might be involved.

**Comment by FabienRoger** [comment]
There is still an issue determining block size with the newlines that isn't present in the ipynb files I tested it with. I will try to fix that latter.

**Review by MathisFederico — DISMISSED** [review]

**Review by vanyle — APPROVED** [review]

**Review by vanyle — APPROVED** [review]

**Pull request merged**
